File: examples/3D/bin/test_3d.js
class ObjMeta  {
  constructor() {
    this.name = "";
  }
}
class test_3d  {
  constructor() {
  }
}
/* static JavaSript main routine at the end of the JS file */
function __js_main() {
  const canvas = document.getElementById("renderCanvas");
  const engine = new BABYLON.Engine(canvas, true, {stencil: true});
  const scene = new BABYLON.Scene(engine);
  
  window.addEventListener('resize', function() {
      engine.resize();
  })
        
  const camera = new BABYLON.FreeCamera("cam1", (new BABYLON.Vector3(0, 11, -25)), scene);
  camera.setTarget(new BABYLON.Vector3(0, 0, 0));
  const loader = new BABYLON.AssetsManager(scene);
  const m = loader.addTextureTask("board",  "https://i.imgur.com/ghjDCqQ.png");
  m.onSuccess = (_task)=> {
    (((event) => { 
      new BABYLON.HemisphericLight("light1", new BABYLON.Vector3(0, 1, 0), scene);
      new BABYLON.DirectionalLight("light2",new BABYLON.Vector3(-1, -1, -1),scene);
      const s1 = new BABYLON.Mesh.CreateSphere("s", 16.0, 1.0, scene);
      const s2 = new BABYLON.Mesh.CreateSphere("s", 16.0, 1.0, scene);
      const hl_layer = new BABYLON.HighlightLayer("hl1", scene);
      const dice = BABYLON.MeshBuilder.CreateBox("b", {width:1.5, height:1.5, depth:1.5, faceUV:[(new BABYLON.Vector4(0.0, 0.0,0.33,0.5)), (new BABYLON.Vector4(0.33, 0.0,0.66,0.5)), (new BABYLON.Vector4(0.66, 0.0,1.0,0.5)), (new BABYLON.Vector4(0.0, 0.5,0.33,1.0)), (new BABYLON.Vector4(0.33, 0.5,0.66,1.0)), (new BABYLON.Vector4(0.66, 0.5,1.0,1.0))]}, scene);
      dice.translate(new BABYLON.Vector3(-4, 5, 4), 1, BABYLON.Space.WORLD ) ;
      dice.rotationQuaternion = new BABYLON.Quaternion.RotationAxis((new BABYLON.Vector3(-1.0, 1.0,0.0)), ((Math.PI) * 0.6));
      const dData = new ObjMeta();
      dice.data = dData;
      dData.name = "dice";
      const points = [new BABYLON.Vector3(0, -2, 0), new BABYLON.Vector3(0, 1, 0)];
      const myTube = BABYLON.MeshBuilder.CreateTube("tube", {path: points, radiusFunction:((i, distance) => { 
        if ( i == 0 ) {
          return 1.0;
        }
        return 0.2;
      }), cap:BABYLON.Mesh.CAP_ALL}, scene);
      const head = new BABYLON.Mesh.CreateSphere("head", 16.0, 1.5, scene);
      head.translate(new BABYLON.Vector3(0, 1, 0), 1, BABYLON.Space.WORLD ) ;
      /** unused:  let merkit = []   **/ 
      const nappula = BABYLON.Mesh.MergeMeshes([myTube, head]);
      nappula.translate(new BABYLON.Vector3(6, 0, 4), 1, BABYLON.Space.WORLD ) ;
      const groundObj = BABYLON.MeshBuilder.CreateBox("board", {width:30.0, height:0.5, depth:30.0}, scene);
      groundObj.translate(new BABYLON.Vector3(-1, -1, -1), 1, BABYLON.Space.WORLD ) ;
      groundObj.data = new ObjMeta();
      s1.data = new ObjMeta();
      s2.data = new ObjMeta();
      const hlMaterial = new BABYLON.StandardMaterial("red", scene);
      hlMaterial.diffuseColor = new BABYLON.Color3(1.0, 0.3, 0.4);
      hlMaterial.specularColor  = new BABYLON.Color3(0.8, 0.2, 1.0);
      hlMaterial.emissiveColor   = new BABYLON.Color3(0.8, 0.2, 0.2);
      hlMaterial.alpha = 0.5;
      const orangeMaterial = new BABYLON.StandardMaterial("red", scene);
      orangeMaterial.diffuseColor = new BABYLON.Color3(1.0, 1.0, 0.4);
      /** unused:  const text2 = new BABYLON.Texture("bump_normal.png", scene)   **/ 
      const redMaterial = new BABYLON.StandardMaterial("red", scene);
      redMaterial.diffuseColor = new BABYLON.Color3(1.0, 0.3, 0.4);
      redMaterial.specularColor  = new BABYLON.Color3(0.8, 0.2, 1.0);
      redMaterial.emissiveColor   = new BABYLON.Color3(0.8, 0.2, 0.2);
      dice.material  = redMaterial;
      nappula.material  = redMaterial;
      nappula.scaling = new BABYLON.Vector3(0.3, 0.3,0.3);
      const n2 = nappula.clone("");
      n2.translate(new BABYLON.Vector3(1, 0, -7), 1, BABYLON.Space.WORLD ) ;
      const n3 = nappula.clone("");
      n3.translate(new BABYLON.Vector3(-4, 0, 3), 1, BABYLON.Space.WORLD ) ;
      const groundMaterial = new BABYLON.StandardMaterial("ground", scene);
      groundMaterial.diffuseColor = new BABYLON.Color3(0.2, 0.3, 0.4);
      groundMaterial.specularColor  = new BABYLON.Color3(1.0, 1.0, 1.0);
      const text = ( (c,g)=>{ var t = new BABYLON.GrassProceduralTexture( "grass" , 1024, scene);t.grassColor = c; t.groundColor=g; return t;})((new BABYLON.Color3(0.2, 0.8, 0.2)), (new BABYLON.Color3(0.2, 0.5, 0.2)));
      text.vScale = -1.0;
      text.uScale = -1.0;
      groundMaterial.diffuseTexture = text;
      const cardMaterial = new BABYLON.StandardMaterial("cardmat", scene);
      cardMaterial.diffuseColor = new BABYLON.Color3(0.2, 0.3, 0.4);
      cardMaterial.specularColor  = new BABYLON.Color3(1.0, 0.3, 0.6);
      const text_1 = new BABYLON.Texture("https://i.imgur.com/ghjDCqQ.png", scene);
      text_1.vScale = -1.0;
      text_1.uScale = -1.0;
      cardMaterial.diffuseTexture = text_1;
      const fancyMaterial = new BABYLON.StandardMaterial("ground", scene);
      fancyMaterial.diffuseColor = new BABYLON.Color3(0.8, 0.41, 0.28);
      fancyMaterial.specularColor  = new BABYLON.Color3(1.0, 1.0, 1.0);
      s1.material  = fancyMaterial;
      s2.material  = fancyMaterial;
      const diceMaterial = new BABYLON.StandardMaterial("dicemat", scene);
      diceMaterial.specularColor  = new BABYLON.Color3(1.0, 0.3, 0.6);
      const text_2 = new BABYLON.Texture("https://i.imgur.com/ghjDCqQ.png", scene);
      text_2.vScale = -1.0;
      text_2.uScale = -1.0;
      diceMaterial.diffuseTexture = text_2;
      dice.material  = diceMaterial;
      s1.translate(new BABYLON.Vector3(2, 0, 0), 1, BABYLON.Space.WORLD ) ;
      s2.translate(new BABYLON.Vector3(-4, 0, -1), 1, BABYLON.Space.WORLD ) ;
      camera.attachControl(canvas, false);
      /** unused:  const zz = -10.0   **/ 
      let time = 0.0;
      let oPos = (s2.position).clone();
      let pickedObj;
      let diceThrow = false;
      let diceStabileCnt = 0;
      canvas.addEventListener("click", (e)=>{ (((ev) => { 
        console.log((("Pointer : " + (scene.pointerX)) + " , ") + (scene.pointerY));
        const pickRes = scene.pick((scene.pointerX), (scene.pointerY));
        const point = pickRes.pickedPoint;
        if ( (typeof(point) !== "undefined" && point != null )  ) {
          const p = point;
          console.log("--> point was : " + (p.x));
          pickedObj = (pickRes.pickedMesh);
          let obj = pickedObj;
          while ((typeof((obj.parent)) !== "undefined" && (obj.parent) != null ) ) {
            obj = (obj.parent);
          };
          pickedObj = obj;
          oPos = (obj.position).clone();
          time = 0.0;
          const d = (pickedObj).data;
          if ( (typeof(d) !== "undefined" && d != null )  ) {
            if( typeof(d) === 'string' ) /* union case for string */ {
              var str = d;
              console.log(str);
            };
          }
          const meta = obj.data;
          if ( (typeof(meta) !== "undefined" && meta != null )  ) {
            if( meta instanceof ObjMeta ) /* union case */ {
              var data = meta;
              if ( (typeof(data.physics) !== "undefined" && data.physics != null )  ) {
                if ( obj == dice ) {
                  diceThrow = true;
                  diceStabileCnt = 10;
                }
                data.physics.applyImpulse(new BABYLON.Vector3(0, 10, 0), p ) ;
              }
            };
          }
        }
      }))(e) } ) 
      scene.enablePhysics();
      const opt1 = {};
      opt1.mass = 0.0;
      opt1.restitution = 0.2;
      const opt2 = {};
      opt2.mass = 1.0;
      opt2.restitution = 0.7;
      opt2.friction = 0.3;
      const opt3 = {};
      opt3.mass = 0.8;
      opt3.restitution = 0.1;
      opt3.friction = 0.4;
      const initGrounds = ((o) => { 
        o.physicsImpostor = new BABYLON.PhysicsImpostor(o, (BABYLON.PhysicsImpostor.BoxImpostor), opt1, scene);
        o.material  = groundMaterial;
      });
      initGrounds(groundObj);
      const applyPhysics = ((obj, opt, type) => { 
        const p_1 = new BABYLON.PhysicsImpostor(obj, type, opt, scene);
        obj.physicsImpostor = p_1;
        let meta_1 = obj.data;
        if ( typeof(meta_1) === "undefined" ) {
          meta_1 = new ObjMeta();
          obj.data = meta_1;
        }
        if ( (typeof(meta_1) !== "undefined" && meta_1 != null )  ) {
          if( meta_1 instanceof ObjMeta ) /* union case */ {
            var data_1 = meta_1;
            data_1.physics = p_1;
          };
        }
      });
      applyPhysics(s1, opt2, BABYLON.PhysicsImpostor.SphereImpostor);
      applyPhysics(s2, opt2, BABYLON.PhysicsImpostor.SphereImpostor);
      applyPhysics(dice, opt2, BABYLON.PhysicsImpostor.BoxImpostor);
      applyPhysics(nappula, opt2, BABYLON.PhysicsImpostor.BoxImpostor);
      applyPhysics(n2, opt2, BABYLON.PhysicsImpostor.BoxImpostor);
      applyPhysics(n3, opt2, BABYLON.PhysicsImpostor.BoxImpostor);
      let all_chips = [];
      const luo_merkki = ((x, y, z, size) => { 
        const merkki = BABYLON.MeshBuilder.CreateTube("tube", {path: [(new BABYLON.Vector3(0.0, -0.1,0.0)), (new BABYLON.Vector3(0.0, 0.1,0.0))], radiusFunction:((i, distance) => { 
          return size * 1.0;
        }), cap:BABYLON.Mesh.CAP_ALL}, scene);
        merkki.translate(new BABYLON.Vector3(x, y,z), 1, BABYLON.Space.WORLD ) ;
        applyPhysics(merkki, opt2, BABYLON.PhysicsImpostor.BoxImpostor);
        merkki.material  = orangeMaterial;
        all_chips.push(merkki);
      });
      luo_merkki(3.0, 3.0, 3.0, 1.0);
      luo_merkki(3.0, 3.3, 3.0, 1.0);
      luo_merkki(3.0, 3.6, 3.0, 1.0);
      luo_merkki(-3.0, 3.0, 3.0, 1.0);
      luo_merkki(-3.0, 3.3, 3.0, 1.0);
      luo_merkki(-3.0, 3.6, 3.0, 1.0);
      const create_card = ((x, y, z) => { 
        const card = BABYLON.MeshBuilder.CreateBox("card", {width:3.0, height:0.1, depth:5.0}, scene);
        card.translate(new BABYLON.Vector3(x, z,y), 1, BABYLON.Space.WORLD ) ;
        card.data = new ObjMeta();
        applyPhysics(card, opt3, BABYLON.PhysicsImpostor.BoxImpostor);
        card.material  = cardMaterial;
        card.rotationQuaternion = new BABYLON.Quaternion.RotationAxis((new BABYLON.Vector3(1.0, 1.0,1.0)), (0.5 * (Math.sin(z))));
      });
      const create_chips = ((n) => { 
        let cnt = n;
        while (cnt > 0) {
          const new_c = ((cnt) => { 
            setTimeout( () => {luo_merkki(3.0, 10.0, 3.0 + (0.1 * (cnt)), 1.0);
            } , 100 * cnt);
          });
          new_c(cnt);
          cnt = cnt - 1;
        };
      });
      let cnt_1 = 10;
      while (cnt_1 > 0) {
        const new_card = ((cnt) => { 
          setTimeout( () => {const rn = Math.sin(((cnt) * 0.1));
          const x = -4.3 + ((cnt) * 0.3);
          /** unused:  const y = -4.3 + rn   **/ 
          create_card(x, -6.3, 11.0 + rn);
          create_card(x, -6.3, 11.0 + rn);
          } , 100 * cnt);
        });
        new_card(cnt_1);
        cnt_1 = cnt_1 - 1;
      };
      /** unused:  const rot = 0.0   **/ 
      const direction = 1.0;
      let last_i = 0.0;
      let last_j = 0.0;
      let last_k = 0.0;
      let highlightCnt = 0;
      engine.runRenderLoop(()=>{
        ((() => { 
          time = time + (0.01 * direction);
          if ( diceStabileCnt > 0 ) {
            const WM = dice.getWorldMatrix();
            const i = WM.m[0 * 4 + 1];
            const j = WM.m[1 * 4 + 1];
            const k = WM.m[2 * 4 + 1];
            if ( (((Math.abs((i - last_i))) < 0.01) && ((Math.abs((j - last_j))) < 0.01)) && ((Math.abs((k - last_k))) < 0.01) ) {
              diceStabileCnt = diceStabileCnt - 1;
            }
            last_i = i;
            last_j = j;
            last_k = k;
            if ( diceStabileCnt == 0 ) {
              const diffAmount = 0.3;
              if ( diceThrow ) {
                const getDiceValue = (() => { 
                  if ( (((Math.abs((1.0 - i))) < diffAmount) && ((Math.abs((0.0 - j))) < diffAmount)) && ((Math.abs((0.0 - k))) < diffAmount) ) {
                    return 1;
                  }
                  if ( (((Math.abs((0.0 - i))) < diffAmount) && ((Math.abs((1.0 - j))) < diffAmount)) && ((Math.abs((0.0 - k))) < diffAmount) ) {
                    return 5;
                  }
                  if ( (((Math.abs((0.0 - i))) < diffAmount) && ((Math.abs((0.0 - j))) < diffAmount)) && ((Math.abs((1.0 - k))) < diffAmount) ) {
                    return 3;
                  }
                  if ( (((Math.abs((-1.0 - i))) < diffAmount) && ((Math.abs((0.0 - j))) < diffAmount)) && ((Math.abs((0.0 - k))) < diffAmount) ) {
                    return 6;
                  }
                  if ( (((Math.abs((0.0 - i))) < diffAmount) && ((Math.abs((-1.0 - j))) < diffAmount)) && ((Math.abs((0.0 - k))) < diffAmount) ) {
                    return 4;
                  }
                  if ( (((Math.abs((0.0 - i))) < diffAmount) && ((Math.abs((0.0 - j))) < diffAmount)) && ((Math.abs((-1.0 - k))) < diffAmount) ) {
                    return 2;
                  }
                  return 0;
                });
                const value = getDiceValue();
                if ( (value > 0) && (value < 6) ) {
                  create_chips(value);
                  switch (value ) { 
                    case 1 : 
                      hl_layer.addMesh(dice, new BABYLON.Color3(0.0, 0.8, 0.0));
                      break;
                    case 2 : 
                      hl_layer.addMesh(dice, new BABYLON.Color3(0.4, 0.8, 0.0));
                      break;
                    case 3 : 
                      hl_layer.addMesh(dice, new BABYLON.Color3(0.4, 0.4, 0.9));
                      break;
                    case 4 : 
                      hl_layer.addMesh(dice, new BABYLON.Color3(0.9, 0.4, 0.9));
                      break;
                    case 5 : 
                      hl_layer.addMesh(dice, new BABYLON.Color3(0.9, 0.9, 0.4));
                      break;
                  };
                  highlightCnt = 50;
                }
                if ( value == 6 ) {
                  hl_layer.addMesh(dice, new BABYLON.Color3(1.0, 0.0, 0.0));
                  highlightCnt = 100;
                  all_chips.forEach(((item, index) => { 
                    const meta_2 = item.data;
                    if ( (typeof(meta_2) !== "undefined" && meta_2 != null )  ) {
                      if( meta_2 instanceof ObjMeta ) /* union case */ {
                        var data_2 = meta_2;
                        if ( (typeof(data_2.physics) !== "undefined" && data_2.physics != null )  ) {
                          data_2.physics.applyImpulse(new BABYLON.Vector3(0, 30, 0), (item.getAbsolutePosition()).add((new BABYLON.Vector3(0.1, 0.1,0.1))) ) ;
                        }
                      };
                    }
                  }));
                }
                diceThrow = false;
              }
            }
          }
          if ( highlightCnt > 0 ) {
            highlightCnt = highlightCnt - 1;
            if ( highlightCnt == 0 ) {
              hl_layer.removeMesh(dice);
            }
          }
          scene.render();
        })) () 
       });
    }))(_task)
      }
      loader.load();
    }
    __js_main();

File: examples/3D/test_3d.clj

Import "Timers.clj"
Import "Engine3D.clj"

class ObjMeta {
  def name ""
  def physics:PhysicsObject3D
}

operator type:void es6 {
  fn log (item:Quaternion3D) ('console.log(' (e 1)')')
}

class test_3d {

  static fn main () {

    ; NOTE: init requires currently HTML having the <canvas id="renderCanvas"></canvas>
    def canvas (getMainCanvas 'renderCanvas')
    def engine (createEngine3D canvas)
    def scene (createScene engine)

    listenToResize engine

    def camera ( newCamera scene (vec 0 11 -25) 'cam1' )
    setTarget camera (vec 0 0 0)

    def loader (assetLoader scene)
    def m ( textureLoader loader 'board' 'https://i.imgur.com/ghjDCqQ.png')
    onReady m {

      newHemisphericLight scene (vec 0 1 0) 'light1'
      newDirectionalLight scene (vec -1 -1 -1) 'light2'
      def s1 (newSphere scene 16.0 1.0 's')
      def s2 (newSphere scene 16.0 1.0 's')
      def hl_layer ( highlight scene 'hl1')

      def dice (box scene 1.5 1.5 1.5 ([] 
          (vec 0.0 0.0 0.33 0.5)
          (vec 0.33 0.0 0.66 0.5)
          (vec 0.66 0.0 1.00 0.5)
          (vec 0.0 0.5 0.33 1.0)
          (vec 0.33 0.5 0.66 1.0)
          (vec 0.66 0.5 1.00 1.0)
        ) 'b')

      translate dice (vec -4 5 4)
      setRotation dice (quat (vec -1.0 1.0 0.0) (M_PI * 0.6) ) 
      def dData (new ObjMeta)
      setData dice dData
      dData.name = 'dice'

      def points ([] (vec 0 -2 0) (vec 0 1 0))
      def myTube (tube scene points {
        if(i == 0 ) {
          return 1.0
        }
        return 0.2
      })
      def head (newSphere scene 16.0 1.5 'head')
      translate head (vec 0 1 0)

      def merkit:[Object3D]
      def nappula (merge ([] myTube head) )
      translate nappula (vec 6 0 4)
      def groundObj (box scene 30.0 0.5 30.0 'board')
      translate groundObj (vec -1 -1 -1)
      setData groundObj (new ObjMeta)
      setData s1 (new ObjMeta)
      setData s2 (new ObjMeta)

      def hlMaterial (createMaterial scene 'red')
      setDiffuse hlMaterial (color 1.0 0.3 0.4)
      setSpecular hlMaterial (color 0.8 0.2 1.0)
      setEmissive hlMaterial (color 0.8 0.2 0.2)
      setAlpha hlMaterial 0.5
  ;     setMaterial highligh hlMaterial

      def orangeMaterial (createMaterial scene 'red')
      setDiffuse orangeMaterial (color 1.0 1.0 0.4)
  ;    setSpecular orangeMaterial (color 0.3 0.1 1.0)

      def text2 (createTexture scene 'bump_normal.png')
      

      def redMaterial (createMaterial scene 'red')
      setDiffuse redMaterial (color 1.0 0.3 0.4)
      setSpecular redMaterial (color 0.8 0.2 1.0)
      setEmissive redMaterial (color 0.8 0.2 0.2)
      setMaterial dice redMaterial

      setMaterial nappula redMaterial
      scale nappula (vec 0.3 0.3 0.3)
      def n2 (clone nappula '')
      translate n2 (vec 1 0 -7 )
      def n3 (clone nappula '')
      translate n3 (vec -4 0 3 )

      def groundMaterial (createMaterial scene 'ground')
      setDiffuse groundMaterial (color 0.2 0.3 0.4)
      setSpecular groundMaterial (color 1.0 1.0 1.0)
      def text (grassTexture scene  (color 0.2 0.8. 0.2) (color 0.2 0.5. 0.2) 1024 )
      vScale text -1.0
      uScale text -1.0
      setTexture groundMaterial text    

      def cardMaterial (createMaterial scene 'cardmat')
      setDiffuse cardMaterial (color 0.2 0.3 0.4)
      setSpecular cardMaterial (color 1.0 0.3 0.6)
      def text (createTexture scene 'https://i.imgur.com/ghjDCqQ.png' )
      vScale text -1.0
      uScale text -1.0
      setTexture cardMaterial text

      def fancyMaterial (createMaterial scene 'ground')
      setDiffuse fancyMaterial (color 0.80 0.41 0.28)
      setSpecular fancyMaterial (color 1.0 1.0 1.0)

      setMaterial s1 fancyMaterial
      setMaterial s2 fancyMaterial

      def diceMaterial (createMaterial scene 'dicemat')
      ; setDiffuse diceMaterial (color 0.2 0.3 0.4)
      setSpecular diceMaterial (color 1.0 0.3 0.6)
      def text (createTexture scene 'https://i.imgur.com/ghjDCqQ.png' )
      vScale text -1.0
      uScale text -1.0
      setTexture diceMaterial text
      setMaterial dice diceMaterial

      translate s1 (vec 2 0 0 )
      translate s2 (vec -4 0 -1 )

      setMouseControl camera canvas

      def zz -10.0
      def time 0.0
      def oPos (clone (getPosition s2))

      def pickedObj:Object3D
      def diceThrow false
      def diceStabileCnt 0

      onClick canvas {
        print "Pointer : " + (getPointerX scene) + " , " + (getPointerY scene)
        def pickRes (pick scene (getPointerX scene) (getPointerY scene))
        def point (hitPoint pickRes)
        if(!null? point) {
          def p (unwrap point)
          print "--> point was : "  + (x p) 
          pickedObj = (unwrap (hitObject pickRes))
          def obj (unwrap pickedObj)
          while( (!null? (parent obj))) {
            obj = (unwrap (parent obj))
          }
          pickedObj = obj
          oPos = (clone (getPosition obj))
          time = 0.0
          def d (getData (unwrap pickedObj))
          if(!null? d) {
            case (unwrap d) str:string {
              print str
            }
          }
          def meta (getData obj)
          if(!null? meta) {
            case (unwrap meta) data:ObjMeta {
              if(!null? data.physics) {
                if( obj == dice ) {
                  diceThrow = true
                  diceStabileCnt = 10
                }
                impulse (unwrap data.physics) (vec 0 10 0) p
              }
            } 
          }
        }
      }
        

      enablePhysics scene

      def opt1 (physicsOptions)
      mass opt1 0.0
      restitution opt1 0.2

      def opt2 (physicsOptions)
      mass opt2 1.0
      restitution opt2 0.7
      friction opt2 0.3

      def opt3 (physicsOptions)
      mass opt3 0.8
      restitution opt3 0.1
      friction opt3 0.4

      def initGrounds (fn:void (o:Object3D) {
        setPhysics o ( physics scene o (physicsBox) opt1 )
        setMaterial o groundMaterial
      })
      initGrounds( groundObj )

      def applyPhysics (fn:void (obj:Object3D opt:PhysicsOptions3D type:PhysicsShape3D) {
        def p ( physics scene obj type opt )
        setPhysics obj p
        def meta (getData obj)
        if(null? meta) {
          meta = (new ObjMeta)
          setData obj (unwrap meta)
        }
        if(!null? meta) {
          case (unwrap meta) data:ObjMeta {
            data.physics = p
          }
        }
      })

      applyPhysics( s1 opt2 (physicsSphere))
      applyPhysics( s2 opt2 (physicsSphere))
      applyPhysics( dice opt2 (physicsBox))

      applyPhysics( nappula opt2 (physicsBox))
      applyPhysics( n2 opt2 (physicsBox))
      applyPhysics( n3 opt2 (physicsBox))

      def all_chips:[Object3D]
      def luo_merkki (fn:void (x:double y:double z:double size:double) {
        def merkki (tube scene ([] (vec 0.0 -0.1 0.0) (vec 0.0 0.1 0.0) ) {
          return size * 1.0
        })
        translate merkki (vec x y z)
        applyPhysics( merkki opt2 (physicsBox))
        setMaterial merkki orangeMaterial 
        push all_chips merkki
      })

      luo_merkki( 3.0 3.0 3.0 1.0)
      luo_merkki( 3.0 3.3 3.0 1.0)
      luo_merkki( 3.0 3.6 3.0 1.0)

      luo_merkki( -3.0 3.0 3.0 1.0)
      luo_merkki( -3.0 3.3 3.0 1.0)
      luo_merkki( -3.0 3.6 3.0 1.0)

      def create_card (fn:void (x:double y:double z:double) {
        def card (box scene 3.0 0.1 5.0 'card')      
        translate card (vec x z y)
        setData card (new ObjMeta)
        applyPhysics( card opt3 (physicsBox))
        setMaterial card cardMaterial
        setRotation card (quat (vec 1.0 1.0 1.0) (0.5 * (sin z)) )
      })

      def create_chips (fn:void (n:int) {
        def cnt n
        while( cnt > 0 ) {
          def new_c (fn:void (cnt:int)  {
            setTimeout {
              luo_merkki( 3.0 10.0 ( 3.0 + (0.1 * (to_double cnt) ) ) 1.0)
            } 100*cnt )
          })     
          new_c(cnt)
          cnt = cnt - 1
        }        
      })

      
      def cnt 10
      while( cnt > 0 ) {
        def new_card (fn:void (cnt:int)  {
          setTimeout {
            def rn (sin (to_double cnt ) * 0.1)
            def x ( -4.3 + (to_double cnt ) * 0.3 )
            def y ( -4.3 + rn )
            create_card( x -6.3 11.0 + rn)
            create_card( x -6.3 11.0 + rn)
          } 100*cnt )
        })     
        new_card(cnt)
        cnt = cnt - 1
      }

      def rot 0.0
      def direction 1.0
      
      def last_i 0.0
      def last_j 0.0
      def last_k 0.0

      def highlightCnt 0

      renderLoop engine {
        time = time + 0.01 * direction
        if( diceStabileCnt > 0 ) {
          def WM (getWorldMatrix dice)
          def i (cell WM 0 1)
          def j (cell WM 1 1)
          def k (cell WM 2 1)          
          if( (near i last_i 0.01) && (near j last_j 0.01) && (near k last_k 0.01) ) {
            diceStabileCnt = diceStabileCnt - 1
          } 
          last_i = i
          last_j = j
          last_k = k

          
          if( diceStabileCnt == 0 ) {
            ; nopan heiton suuruuden määritys...
            def diffAmount 0.3
            if diceThrow {
              def getDiceValue (fn:int () {
                if( (near 1.0 i diffAmount) && (near 0.0 j diffAmount) && (near 0.0 k diffAmount)) {
                  return 1
                }
                if( (near 0.0 i diffAmount) && (near 1.0 j diffAmount) && (near 0.0 k diffAmount)) {
                  return 5
                }
                if( (near 0.0 i diffAmount) && (near 0.0 j diffAmount) && (near 1.0 k diffAmount)) {
                  return 3
                }
                if( (near -1.0 i diffAmount) && (near 0.0 j diffAmount) && (near 0.0 k diffAmount)) {
                  return 6
                }
                if( (near 0.0 i diffAmount) && (near -1.0 j diffAmount) && (near 0.0 k diffAmount)) {
                  return 4
                }
                if( (near 0.0 i diffAmount) && (near 0.0 j diffAmount) && (near -1.0 k diffAmount)) {
                  return 2
                }
                return 0
              })
              def value (getDiceValue())
              if( value > 0 &&  value < 6 ) {
                create_chips( value )
                switch value {
                  case 1 {
                    add hl_layer dice (color 0.0 0.80 0.0)
                  }
                  case 2 {
                    add hl_layer dice (color 0.4 0.80 0.0)
                  }
                  case 3 {
                    add hl_layer dice (color 0.4 0.40 0.9)
                  }
                  case 4 {
                    add hl_layer dice (color 0.9 0.40 0.9)
                  }
                  case 5 {
                    add hl_layer dice (color 0.9 0.90 0.4)
                  }
                }
                highlightCnt = 50
              } 
              if( value == 6) {
                add hl_layer dice (color 1.0 0.0 0.0)
                highlightCnt = 100
                forEach all_chips {
                  def meta (getData item)
                  if(!null? meta) {
                    case (unwrap meta) data:ObjMeta {
                      if(!null? data.physics) {
                        impulse (unwrap data.physics) (vec 0 30 0) ( (getAbsolutePosition item) + (vec 0.1 0.1 0.1))   
                      }
                    }
                  }                                                    
                }
                
              }
              diceThrow = false
            }
          } 

        }        
        if( highlightCnt > 0 ) {
          highlightCnt = highlightCnt - 1
          if( highlightCnt == 0) {
            remove hl_layer dice            
          }
        }
        render scene
      }
    }
    start loader
  }
}